### ISSUE TITLE
Add support for Delta-CRDT

### ISSUE THREAD

**Issue by mboogerd** [akka/akka#21875]
The paper "Delta State Replicated Data Types" from Paulo, Shoker, Baquero describes a number of Delta-CRDTs which are more efficient in their state-dissemination:

- Grow-only set
- 2P (two-phase) set
- Add-wins last-writer-wins set
- Positive-Negative counter
- Lexicographic counter
- Enable-wins flag
- Multi-value register
- Add-wins set
- Remove-wins set
- Observed-removed map

I have implemented the last five in a separate project and I can take a shot at porting them to Akka distributed-data. Some questions regarding this:

1) Would there be interest in the first five?
2) I have developed a directed-graph delta-CRDT, and will be implementing a hyper-graph as well as, by lack of an official term, "flow-graph". The latter is a hyper-graph with special sets for source/sink vertices, i.e. vertices that can only initiate/terminate directed edges, respectively. My use-case is eventual-consistent data-flow definitions. I imagine this could be of interest, just let me know.
3) A pickle I should have realised, but didn't; As I initially didn't plan on adding these to Akka, my implementations rely on algebra, in particular [BoundedJoinSemilattice](https://github.com/typelevel/algebra/blob/master/core/src/main/scala/algebra/lattice/BoundedJoinSemilattice.scala). Also, I rely on typeclasses in general to derive valid compositions of lattices. Just to be sure, I'm guessing I have to substitute the FP approach with OO to support Java...?

**Comment by patriknw** [comment]
This is very interesting, @mboogerd 
I have read the delta-CRDT paper several times and think it would be excellent to add support for that in Akka Distributed Data (DData), but I have never got around to implementing any of it. Looking forward to review your code and guide you.

> first five

Currently we have GSet and PNCounter, so those are interesting as delta-CRDTs also. Lexicographic counter sounds interesting.

Before adding more types I would like to see a small prototype of how it would integrate with DData. We are fine with changing the APIs if needed. It's still experimental, but we must make the APIs stable in Akka 2.5, which is about to happen the next few months, so we should first focus on what might be needed in the API to support delta-CRDTs. The actual implementation doesn't have to be included in 2.5.

> hyper-graph

Everything is interesting, but let's do it step-by-step.

> to support Java

Yes, we must have Java and Scala API for all features.

**Comment by mboogerd** [comment]
@patriknw, you're right, getting delta-CRDTs to work in DD should be the first challenge. I'll try to create something that is useful to review in the next few weeks.

**Comment by patriknw** [comment]
thanks, just ping me if you have any questions

**Comment by mboogerd** [comment]
@patriknw: After having looked at the code, my first assessment in descending order of priority:

**Pruning**
Imo, this is the only part of distributed data that is not a straightforward read (kudos really, everything else looks very tidy :)); it complicates the design of both the collection CRDTs and the replicator. Unbounded growth of replica-ids can be a problem, but:
- Given that message size of Delta-CRDTs contains only identifiers of replicas relevant to the delta, the unbounded growth plays no significant factor there.
- Unbounded growth of the state at replicas does not have to be a (memory) issue if we introduce persistence. It requires no complicated coordination, does not limit scalability and is on itself beneficial. I would feel it is the better investment (on that note: are there any thoughts/preferences within the Akka team/users for CRDT persistence?)
- If necessary, a process of replica-id-cleanup could be embedded inside CRDTs. For example, one can maintain a mapping of pruned-dots to dots-of-dummy-substituting-replica, together with a version counter for prune operations. The mapping can be pruned after a bounded number of pruning-operations. Up to that point, older delta's can be transformed. When encountering delta's with a pruning-version below that bound one has the choice making the merge effective or ineffective. Either way, it should prevent unbounded growth.

I would like to hear your thoughts on the above; if we agree that pruning should be solved differently in the future, then I won't focus on implementing pruning for delta-CRDTs

**Replication**
The referenced paper assumes an atomic broadcast of delta's for the sake of simplicity. DD uses a gossip-protocol with read-repair and quorum/majority/all-write. This poses us with some design decision:

1. Implement an atomic broadcast protocol. As these protocols require coordination, they are less scalable than gossip. Also, I think their consistency properties may be stronger than what we require.
2. Maintaining the current protocol with minimal adjustment. As delivery and its order are not guaranteed, newer deltas may be received before older deltas. We have two viable options:
2.1. Eventual consistency: Merging the delta and optionally initiating anti-entropy (we have evidence that more information exists, so it seems prudent), or
2.2. Causal consistency: Postponing the merge, and initiate anti-entropy such that delta's can be merged in causal order
3. On a related note, we can choose to have replicas only dispatch delta's initiated at their own site, or have them also dispatch delta's received by others.
4. Implement an alternative gossip protocol. There are many interesting choices (plumtree, scuttlebutt, knn-graph-approximations), but to my knowledge few comparisons in recent years. Perhaps within the Akka team/userbase there are already alternatives/preferences?

Again, your opinion on these matters would be appreciated. My take on the matter is:
- Start with 2.1
- Add 2.2 as a configuration parameter on CRDT-level
- Experiment with 3/4.x to find what works well for typical DD workloads

**Comment by mboogerd** [comment]
@patriknw: Here's a gist for the style I'd propose for porting delta-CRDTs to Java-compatible Scala. It retains the composability of lattices (which is important to keep the code DRY, especially with the more complicated lattices), but does not use any implicits to derive lattice-instances for the CRDTs.

https://gist.github.com/mboogerd/28679f28a1495fcfc0ff5f373f2b9fb5

Naturally, the code contains a good amount of comments to make clear what everything does and how it is supposed to be used. Left that out for brevity. If you could just give me a thumbs up or so, that'd be great.

**Comment by patriknw** [comment]
@mboogerd Thanks for good starting points for discussion here. I think your suggestions sounds good.

> Pruning

I agree. I'm not happy with that part. We can absolutely try to find other solutions or conclude that it's not needed. Especially since I recently found some issues with the implementation, see https://github.com/akka/akka/issues/21647

> Unbounded growth of the state at replicas does not have to be a (memory) issue if we introduce persistence.

In what why would persistence solve the issue. I would say that persistence makes the issue even worse, since such node identifiers are never removed (without persistence they are at least removed when the whole cluster is shutdown).

I have actually an implementation for persistence in progress here https://github.com/akka/akka/pull/21646
It would be great if you can review and think about how delta-crdts would influence the requirements for the durable store.

> Replication

First, do you think we should implement the gossip protocol for causal consistency that is described in the paper?

I'll probably have more comments later

**Comment by mboogerd** [comment]
> In what why would persistence solve the issue. I would say that persistence makes the issue even worse, since such node identifiers are never removed (without persistence they are at least removed when the whole cluster is shutdown).

You're right, persistence doesn't solve anything with respect to the unbounded growth. Persistent storage simply is some orders cheaper, making the problem less costly. In fact, I suspect it is cheap enough to prevent it from being an issue for the majority of cases. But of course, I could be wrong there, as I have little practical experience.

> First, do you think we should implement the gossip protocol for causal consistency that is described in the paper?

My previous points 1 (atomic-broadcast) and 3 (direct vs. transitive-mode) are essentially the extent of the protocol-specification provided by the paper. This means we are left with a choice of topology. This could range from having each replica atomic-broadcast its mutations to all other replicas in the cluster, do random one-on-one rounds of gossip, or anything in between (such as epidemic broadcast trees). The latter two may need some repair mechanisms (read-repair, anti-entropy, hinted-handoff). I'm not sure whether there is one right choice for Akka as DData may be used for various purposes. If so, a configurable replication-strategy might be a safe bet. What do you think? Also, do you have ideas for a default/general-purpose replication-strategy?

If I have time this weekend, I will do some re-reading of papers. My focus will be [Plumtree (Epidemic Broadcast Trees)](https://www.google.nl/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwij84elhcbQAhXImBoKHREwAvIQFggdMAA&url=http%3A%2F%2Fwww.gsd.inesc-id.pt%2F~jleitao%2Fpdf%2Fsrds07-leitao.pdf&usg=AFQjCNGo4mQt7dU0A5Qgt6US90iyzwWo4g), [Scuttlebutt](https://www.google.nl/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwjbnrfPhcbQAhXEbBoKHXU2CQYQFgggMAA&url=https%3A%2F%2Fwww.cs.cornell.edu%2Fhome%2Frvr%2Fpapers%2Fflowgossip.pdf&usg=AFQjCNGIo_g2UH39zMvtuAGVGd_Q8iRdiA), [Gossip Kit](https://www.google.nl/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwj0nO-liMbQAhUFfRoKHSp8DPMQFggfMAA&url=https%3A%2F%2Fhal.inria.fr%2Fhal-01080198%2Ffile%2FTSE-GossipKit.pdf&usg=AFQjCNFp3eD5QywJDUlJq6VXvecJb6pCSw). That would be my current selection for an initial experimental/configurable implementation. Do you have any additional sources that could be inspirational for this process?

> I have actually an implementation for persistence in progress here #21646
It would be great if you can review and think about how delta-crdts would influence the requirements for the durable store.

I will have a look on your persistence implementation, very curious towards the direction you have chosen!

**Comment by patriknw** [comment]
> unbounded growth

Let's say that we have a GCounter with 10 new participating nodes per day (in an elastic cluster nodes are added and removed). Currently the node identifier is a `UniqueAddress`. I have not checked what size a typical address has, but let's say that it's 100 bytes. That means that after one year the GCounter has grown to 10 * 365 * 100 = 265,000 bytes.

We could reduce the impact by using a more compact node identifier, such as a Long (8 bytes). In the above example that means 29,200 bytes per year. Can that be ignored? 

> Replication

I don't mean that we need to find an advanced/better dissemination algorithm. Random gossip as in current implementation is probably just fine.

I'm curious about the causal consistency aspects when using delta-CRDTs.
Without an algorithm that enforces casual ordering when applying the deltas it's possible to see updates in "wrong" order. E.g. GSet {}, add A, add B => on other nodes that can be seen as {B}. With causal ordering it can only be seen as {A} or {A, B}. 

To implement the causal merging we must keep track of sequence numbers of the delta-intervals, as described in "Anti-Entropy Algorithm for Causal Consistency" in the paper.

We can of course do it in two steps, and start with only eventual consistency, as described in the "Basic Anti-Entropy Algorithm" in the paper.

We must also keep in mind that DData has two ways to replicate the data.
1. the immediate replication as specified in the `WriteConsistency` of the `Update`
2. the background gossip dissemination

**Comment by patriknw** [comment]
Let's continue discussion of unbounded growth in https://github.com/akka/akka/issues/21647 and not worry too much about that here.

**Comment by mboogerd** [comment]
Apologies for my neglect of this issue recently. I started experimenting with a minimal rewrite of the current implementation as well as rewriting it from scratch (to improve my understanding of DData's consistency levels in combination with delta-CRDTs), but I have been mostly occupied otherwise. I did check your persistence implementation, but it's too early for me to tell whether it is going to work with delta-CRDTs. Lousy expectation management is the best I have to offer, which is that I feel the inspirational surge, but there will be limited opportunities this month. I guess this is more letting you know that I haven't completely abandoned it, and I will spend a few hours this weekend.
